packages/caver-wallet/src/keyring/keyring.js

```diff
@@ -0,0 +1,845 @@
+
+        if (chainId === undefined) {
+            throw new Error(`chainId should be defined to sign.`)
+        } else {
+            chainId = utils.toHex(chainId)
+        }
+        if (role === undefined) throw new Error(`role should be defined to sign.`)
+
+        const keys = this.getKeyByRole(role)
+        if (index >= keys.length) throw new Error(`Invalid index(${index}): index must be less than the length of keys(${keys.length}).`)
+        return keys[index].sign(transactionHash, chainId)
+    }
+
+    /**
+     * signs with transactionHash with multiple keys and returns signatures.
+     *
+     * @param {string} transactionHash The hash of transaction.
+     * @param {string|number} chainId The chainId or the network.
+     * @param {number} role A number indicating the role of the key. You can use `caver.wallet.keyring.role`.
+     * @return {Array.<Array<string>>}
+     */
+    signWithKeys(transactionHash, chainId, role) {
+        if (!utils.isTxHashStrict(transactionHash)) throw new Error(`Invalid transaction hash: ${transactionHash}`)
+
+        if (chainId === undefined) {
+            throw new Error(`chainId should be defined to sign.`)
+        } else {
+            chainId = utils.toHex(chainId)
+        }
+        if (role === undefined) throw new Error(`role should be defined to sign.`)
+
+        const signatures = []
+
+        for (const key of this.getKeyByRole(role)) {
+            signatures.push(key.sign(transactionHash, chainId))
+        }
+
+        return signatures
+    }
+
+    /**
+     * signs with hashed data and returns result object that includes `signsture`, `message` and `messageHash`
+     *
+     * @param {string} data The data string to sign.
+     * @param {number} [role] A number indicating the role of the key. You can use `caver.wallet.keyring.role`. (default: `caver.wallet.keyring.role.ROLE_TRANSACTION_KEY`)
+     * @param {number} [index] The index of the key to be used. (default: 0)
+     * @return {object}
+     */
+    signMessage(data, role, index) {
+        const messageHash = utils.hashMessage(data)
+        if (role === undefined && index === undefined) {
+            role = KEY_ROLE.ROLE_TRANSACTION_KEY
+            if (this._key[role].length === 0) throw new Error(`Dafault key(${KEY_ROLE[0]}) does not have enough key to sign.`)
+            index = 0
+        } else if (role === undefined || index === undefined) {
+            throw new Error(
+                `To sign message, both role and index must be defined. ` +
+                    `If both role and index are not defined, this function signs the message using the default key(${KEY_ROLE[0]}[0]).`
+            )
+        }
+
+        const keys = this.getKeyByRole(role)
+        if (index >= keys.length) throw new Error(`Invalid index(${index}): index must be less than the length of keys(${keys.length}).`)
+
+        const signed = keys[index].signMessage(messageHash)
```

Comment:
Don't we need `signed.messageHash = messageHash`? Is the variable is set in the function?

Reply:
Originally, {signature, messageHash} object is made in PrivateKey.
I modified the logic, thinking privateKey.signMessage would return the signature and constructing the result object from keyring.signMessage is more clear.